apps/src/lib/node/ledger/shell/vote_extensions.rs

```diff
@@ -0,0 +1,460 @@
+        /// signed and prepared for inclusion in a vote extension.
+        pub fn new_ethereum_events(&mut self) -> Vec<SignedEthEvent> {
+            let mut events = vec![];
+            let voting_power = self.get_validator_voting_power();
+            let address = self.mode.get_validator_address().cloned();
+            if let ShellMode::Validator {
+                ref mut ethereum_recv,
+                data:
+                    ValidatorData {
+                        keys:
+                            ValidatorKeys {
+                                protocol_keypair, ..
+                            },
+                        ..
+                    },
+                ..
+            } = &mut self.mode
+            {
+                let (voting_power, address) =
+                    voting_power.zip(address).unwrap();
+                while let Ok(eth_event) = ethereum_recv.try_recv() {
+                    events.push(SignedEthEvent::new(
+                        eth_event,
+                        address.clone(),
+                        voting_power.clone(),
+                        self.storage.last_height + 1,
+                        protocol_keypair,
+                    ));
+                }
+            }
+            events
+        }
+
+        /// Verify that each ethereum header in a vote extension was signed by
+        /// a validator in the correct epoch, the stated voting power is
+        /// correct, and the signature is correct.
+        pub fn validate_ethereum_event(
+            &self,
+            height: BlockHeight,
+            event: &impl SignedEvent,
+        ) -> bool {
+            let epoch = self.storage.block.pred_epochs.get_epoch(height);
+            let total_voting_power = self.get_total_voting_power(epoch);
+            if u64::from(total_voting_power) == 0 {
+                return false;
+            }
+
+            // Get the public keys of each validator. Filter out those that
+            // inaccurately stated their voting power at a given block height
+            let mut public_keys = vec![];
+            for EpochPower {
+                validator,
+                voting_power,
+                block_height,
+            } in event.get_voting_powers().into_iter()
+            {
+                if block_height != height {
+                    continue;
+                }
+                if let Some((power, pk)) =
+                    self.get_validator_from_address(&validator, epoch)
+                {
+                    let power =
+                        FractionalVotingPower::new(power, total_voting_power)
+                            .unwrap();
+                    if power == voting_power {
+                        public_keys.push(pk);
+                    }
+                }
+            }
+            // check that we found all the public keys and
+            // check that the signatures are valid
+            public_keys.len() == event.number_of_signers()
```

Comment:
if we don't find all public keys, this is the validator's error so we should return an error rather than reject the event as invalid. Even if functionally the behaviour will be the same, it'd be clearer in the code I think

Reply:
This method will actually be called by the validator proposing the block to make sure they got everything right as well as others checking the proposer's work. This is also used to weed out malformed vote extensions as a result. I could put an error here, but it would require enlarging the error enum and I would never print the message out anyway.